### PR TITLE
Adds ipv4 and ipv6 formats support

### PR DESCRIPTION
**Pull request by RiccardoManfrin**
trash it if too poor quality

### CODE REVIEW

**Comment by RiccardoManfrin** [comment]
I can rollback to `DefineStringFormat` np. It's gross patch for my stuff to work, didn't care about backward compat, but I agree.
I'd keep the `DefineStrinngCallbackFormat` one instead.. more explicit.. just taste btw.

File: openapi3/schema.go [hunk 1]
```diff
@@ -930,14 +930,25 @@ func (schema *Schema) visitJSONString(value string, fast bool) (err error) {
 			schema.compiledPattern = cp
 		} else if v := schema.Format; len(v) > 0 {
 			// No pattern, but does have a format
-			re := SchemaStringFormats[v]
-			if re != nil {
-				cp = &compiledPattern{
-					Regexp:    re,
-					ErrReason: "JSON string doesn't match the format '" + v + " (regular expression `" + re.String() + "`)'",
+			f := SchemaStringFormats[v]
+			switch f.Type {
```
Comment (fenollp) [inline openapi3/schema.go:934]: Prefer if/else on `f.Regexp == nil` and remove f.Type altogether plz

File: openapi3/schema.go [hunk 2]
```diff
@@ -953,6 +964,13 @@ func (schema *Schema) visitJSONString(value string, fast bool) (err error) {
 				Reason:      cp.ErrReason,
 			}
 		}
+	} else if len(schema.Format) > 0 && cp == nil {
```
Comment (fenollp) [inline openapi3/schema.go:967]: Please open a separate PR for this as
> Tools that do not recognize a specific format MAY default back to the type alone, as if the format is not specified.

I'd prefer this functionality to be disable-able via a visit setting and I'd lean on unknown formats being allowed by default (thoughts about that default?)

File: openapi3/schema_formats.go [hunk 3]
```diff
@@ -10,15 +11,89 @@ const (
...
+
+// FormatTypeCallback is callback based validation
+const FormatTypeCallback = "callback"
+
+type FormatCallback func(Val string) error
+
+//Format is the format type context fo validate the format
+type Format struct {
+	Type     FormatType
+	Regexp   *regexp.Regexp
```
Comment (fenollp) [inline openapi3/schema_formats.go:28]: Let's keep Regexp,Callback fields private

File: openapi3/schema_formats.go [hunk 4]
```diff
@@ -35,4 +110,8 @@ func init() {
 
 	// date-time
 	DefineStringFormat("date-time", `^[0-9]{4}-(0[0-9]|10|11|12)-([0-2][0-9]|30|31)T[0-9]{2}:[0-9]{2}:[0-9]{2}(.[0-9]+)?(Z|(\+|-)[0-9]{2}:[0-9]{2})?$`)
+
+	DefineStringCallbackFormat("ipv4", validateIPv4)
+	DefineStringCallbackFormat("ipv6", validateIPv6)
```
Comment (fenollp) [inline openapi3/schema_formats.go:115]: These are not part of the OAS so should not be predefined. cc https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#data-types

However I am not against
```go
func DefineIPv4Format() { ... }
func DefineIPv6Format() { ... }
```
so user can easily opt in.

**Review by fenollp — CHANGES_REQUESTED** [review]

File: openapi3/schema.go (hunk 2 above)
Comment (fenollp) [reply openapi3/schema.go:961]: Again, this is not per-OpenAPIv3 IMO. Please move this to another PR (and add a new `SchemaValidationOption` "DisallowUnsupportedFormat()" maybe?)

File: openapi3/schema.go [hunk 5]
```diff
@@ -930,13 +930,18 @@ func (schema *Schema) visitJSONString(value string, fast bool) (err error) {
 			schema.compiledPattern = cp
 		} else if v := schema.Format; len(v) > 0 {
 			// No pattern, but does have a format
-			re := SchemaStringFormats[v]
-			if re != nil {
-				cp = &compiledPattern{
-					Regexp:    re,
-					ErrReason: "JSON string doesn't match the format '" + v + " (regular expression `" + re.String() + "`)'",
+			f := SchemaStringFormats[v]
+			if f.regexp != nil {
```
Comment (fenollp) [inline openapi3/schema.go:934]: case switch on
* `f.regexp!=nil && f.callback==nil`
* `f.regexp==nil && f.callback!=nil`
* default: return an error "corrupted entry %q in SchemaStringFormats", v

Also: there needs to be a `f!=nil` check

File: openapi3/schema.go [hunk 6]
```diff
@@ -930,13 +930,18 @@ func (schema *Schema) visitJSONString(value string, fast bool) (err error) {
...
+				re := f.regexp
+				if re != nil {
```
Comment (fenollp) [inline openapi3/schema.go:936]: ...

File: openapi3/schema_formats.go [hunk 7]
```diff
@@ -10,15 +11,86 @@ const (
...
-	SchemaStringFormats[name] = re
+	f := Format{
+		regexp:   re,
+		callback: nil}
+	SchemaStringFormats[name] = f
+}
+
+// DefineStringCallbackFormat define callback based type callback validation
+func DefineStringCallbackFormat(name string, callback FormatCallback) {
+	f := Format{
```
Comment (fenollp) [inline openapi3/schema_formats.go:50]:
```suggestion
	SchemaStringFormats[name] = Format{callback: callback}
```

File: openapi3/schema_formats.go [hunk 8]
```diff
@@ -10,15 +11,86 @@ const (
...
+
+//DefineStringFormat Defines a new regexp pattern for a given format
 func DefineStringFormat(name string, pattern string) {
 	re, err := regexp.Compile(pattern)
 	if err != nil {
 		err := fmt.Errorf("Format '%v' has invalid pattern '%v': %v", name, pattern, err)
 		panic(err)
 	}
-	SchemaStringFormats[name] = re
+	f := Format{
```
Comment (fenollp) [inline openapi3/schema_formats.go:42]:
```suggestion
	SchemaStringFormats[name] = Format{regexp: re}
```

File: openapi3/schema_formats.go [hunk 9]
```diff
@@ -10,15 +11,86 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
 
-var SchemaStringFormats = make(map[string]*regexp.Regexp, 8)
+// FormatType is the type of format validation used
+type FormatType string
```
Comment (fenollp) [inline openapi3/schema_formats.go:15]: delete

File: openapi3/schema_formats.go [hunk 10]
```diff
@@ -10,15 +11,86 @@ const (
...
 
+// FormatTypeRe is regexp based validation
+const FormatTypeRe = "re"
```
Comment (fenollp) [inline openapi3/schema_formats.go:18]: delete

File: openapi3/schema_formats.go [hunk 11]
```diff
@@ -10,15 +11,86 @@ const (
...
+
+// FormatTypeCallback is callback based validation
+const FormatTypeCallback = "callback"
```
Comment (fenollp) [inline openapi3/schema_formats.go:21]: delete

File: openapi3/schema_formats.go [hunk 12]
```diff
@@ -10,15 +11,86 @@ const (
...
+
+//FormatCallback custom check on exotic formats
+type FormatCallback func(Val string) error
+
+//Format is the format type context fo validate the format
```
Comment (fenollp) [inline openapi3/schema_formats.go:26]: try harder

File: openapi3/schema_formats.go [hunk 13]
```diff
@@ -10,15 +11,86 @@ const (
...
 	}
-	SchemaStringFormats[name] = re
+	f := Format{
+		regexp:   re,
+		callback: nil}
+	SchemaStringFormats[name] = f
+}
+
+// DefineStringCallbackFormat define callback based type callback validation
+func DefineStringCallbackFormat(name string, callback FormatCallback) {
```
Comment (fenollp) [inline openapi3/schema_formats.go:49]:
```suggestion
func DefineStringFormatCallback(name string, callback FormatCallback) {
```

File: openapi3/schema_formats.go [hunk 14]
```diff
@@ -10,15 +11,86 @@ const (
...
 		panic(err)
 	}
-	SchemaStringFormats[name] = re
+	f := Format{
+		regexp:   re,
+		callback: nil}
+	SchemaStringFormats[name] = f
+}
+
+// DefineStringCallbackFormat define callback based type callback validation
```
Comment (fenollp) [inline openapi3/schema_formats.go:48]:
```suggestion
// DefineStringFormatCallback ads a validation function for a specific schema format entry
```

File: openapi3/schema_formats.go [hunk 15]
```diff
@@ -35,4 +107,15 @@ func init() {
 
 	// date-time
 	DefineStringFormat("date-time", `^[0-9]{4}-(0[0-9]|10|11|12)-([0-2][0-9]|30|31)T[0-9]{2}:[0-9]{2}:[0-9]{2}(.[0-9]+)?(Z|(\+|-)[0-9]{2}:[0-9]{2})?$`)
+
+}
+
+// DefineIPv4Format opts in ipv4 format validation on top of OAS 3 spec
+func DefineIPv4Format() {
+	DefineStringCallbackFormat("ipv4", validateIPv4)
```
Comment (fenollp) [inline openapi3/schema_formats.go:115]:
```suggestion
	DefineStringFormatCallback("ipv4", validateIPv4)
```

File: openapi3/schema_formats.go [hunk 16]
```diff
@@ -35,4 +107,15 @@ func init() {
...
+}
+
+// DefineIPv6Format opts in ipv6 format validation on top of OAS 3 spec
+func DefineIPv6Format() {
+	DefineStringCallbackFormat("ipv6", validateIPv6)
```
Comment (fenollp) [inline openapi3/schema_formats.go:120]:
```suggestion
	DefineStringFormatCallback("ipv6", validateIPv6)
```

**Review by fenollp — CHANGES_REQUESTED** [review]
I changed to DefineStringFormatCallback so things are easier to grep.

The rest should be pretty straightforward.

File: openapi3/schema.go (hunk 6 above)
Comment (RiccardoManfrin) [reply openapi3/schema.go:936]: :)

File: openapi3/schema_formats.go (hunk 12 above)
Comment (RiccardoManfrin) [reply openapi3/schema_formats.go:26]: you mean on the comments? suggestions?

Comment (fenollp) [reply openapi3/schema_formats.go:26]: Yes I'm not sure what you're trying to convey here. Just delete it?

**Review by fenollp — CHANGES_REQUESTED** [review]
Good job! Still some outstanding comments though.

File: openapi3/schema_formats.go [hunk 17]
```diff
@@ -10,15 +11,80 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
 
-var SchemaStringFormats = make(map[string]*regexp.Regexp, 8)
+// FormatType is the type of format validation used
```
Comment (fenollp) [inline openapi3/schema_formats.go:14]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 18]
```diff
@@ -10,15 +11,80 @@ const (
...
+type FormatType string
```
Comment (fenollp) [inline openapi3/schema_formats.go:15]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 19]
```diff
@@ -10,15 +11,80 @@ const (
...
 
+// FormatTypeRe is regexp based validation
```
Comment (fenollp) [inline openapi3/schema_formats.go:17]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 20]
```diff
@@ -10,15 +11,80 @@ const (
...
+const FormatTypeRe = "re"
```
Comment (fenollp) [inline openapi3/schema_formats.go:18]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 21]
```diff
@@ -10,15 +11,80 @@ const (
...
+
+// FormatTypeCallback is callback based validation
+const FormatTypeCallback = "callback"
+
+//FormatCallback custom check on exotic formats
+type FormatCallback func(Val string) error
+
+//Format is the format type context fo validate the format
```
Comment (fenollp) [inline openapi3/schema_formats.go:26]:
```suggestion
```

**Review by fenollp — CHANGES_REQUESTED** [review]

File: openapi3/schema_formats.go [hunk 22]
```diff
@@ -10,15 +11,75 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
 
-var SchemaStringFormats = make(map[string]*regexp.Regexp, 8)
 
+
+// FormatTypeCallback is callback based validation
+const FormatTypeCallback = "callback"
```
Comment (fenollp) [inline openapi3/schema_formats.go:17]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 23]
```diff
@@ -10,15 +11,75 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
 
-var SchemaStringFormats = make(map[string]*regexp.Regexp, 8)
 
+
+// FormatTypeCallback is callback based validation
```
Comment (fenollp) [inline openapi3/schema_formats.go:16]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 24]
```diff
@@ -10,15 +11,75 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
 
-var SchemaStringFormats = make(map[string]*regexp.Regexp, 8)
 
+
```
Comment (fenollp) [inline openapi3/schema_formats.go:15]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 25]
```diff
@@ -10,15 +11,72 @@ const (
...
```
Comment (fenollp) [inline openapi3/schema_formats.go:15]:
```suggestion
```

Comment (fenollp) [inline openapi3/schema_formats.go:15]:
```suggestion
```

File: openapi3/schema_formats.go [hunk 26]
```diff
@@ -10,15 +11,71 @@ const (
 	FormatOfStringForUUIDOfRFC4122 = `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
 )
```
Comment (fenollp) [inline openapi3/schema_formats.go:13]:
```suggestion
```

**Pull request merged**